### PR TITLE
Add selective weight loading decode kernel for activation sparsity

### PR DESCRIPTION
**Pull request by jcaip**
Summary:

This PR adds in a kernel to accelerate $$xW^T$$ when x is sparse and we are memory bound.

The idea here is that we can avoid loading the columns of $W$ that correspond to the zero elements of $x$.

This lets us accelerate activation sparsity for bs=1 decode use cases.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

### CODE REVIEW

File: torchao/dtypes/floatx/cutlass_semi_sparse_layout.py [hunk 1]
```diff
@@ -15,15 +15,51 @@
...
+from torchao.dtypes.utils import AQTTensorImpl, Layout, get_out_shape
 from torchao.ops import (
     rowwise_scaled_linear_sparse_cutlass_f8f8,
-    to_sparse_semi_structured_cutlass_sm9x_f8,
 )
 
 aten = torch.ops.aten
 
 
+def _pad_dense_input(dense_input: torch.Tensor) -> torch.Tensor:
```
Comment (jerryzh168) [inline torchao/dtypes/floatx/cutlass_semi_sparse_layout.py:26]: I think ideally we don't do padding at this level, since this interferes with aliasing semantics, and make some op implementation impossible, like slice. is it possible to move this to the kernel itself?

File: torchao/dtypes/floatx/cutlass_semi_sparse_layout.py [hunk 2]
```diff
@@ -66,11 +97,12 @@ def __new__(
         )
         kwargs["dtype"] = sparse.dtype
         kwargs["requires_grad"] = False
-        shape = (sparse.shape[0], 2 * sparse.shape[-1])
+        # shape = (sparse.shape[0], 2 * sparse.shape[-1])
```
Comment (jerryzh168) [inline torchao/dtypes/floatx/cutlass_semi_sparse_layout.py:100]: remove?

File: torchao/dtypes/floatx/cutlass_semi_sparse_layout.py [hunk 3]
```diff
@@ -80,6 +112,7 @@ def __init__(
         self.meta = meta
         self.scale = scale
         self._layout = _layout
+        self._shape = shape
```
Comment (jerryzh168) [inline torchao/dtypes/floatx/cutlass_semi_sparse_layout.py:115]: why do we need a separate shape, instead of deriving from the `sparse` tensor?

Comment (jerryzh168) [reply torchao/dtypes/floatx/cutlass_semi_sparse_layout.py:115]: nit: also sparse naming might be a bit vague, renaming to `sparse_data` or something might be better

File: torchao/dtypes/floatx/cutlass_semi_sparse_layout.py [hunk 4]
```diff
@@ -210,3 +257,46 @@ def _linear_fp8_act_fp8_weight_sparse_cutlass_impl(input_tensor, weight_tensor,
     )
 
     return out
+
+
+def _linear_fp8_act_sparse_fp8_weight_cutlass_check(input_tensor, weight_tensor, bias):
+    from torchao.dtypes.floatx import Float8Layout
+
+    res = (
+        isinstance(input_tensor, AffineQuantizedTensor)
```
Comment (jerryzh168) [inline torchao/dtypes/floatx/cutlass_semi_sparse_layout.py:266]: just FYI, I feel we could simplify this list of checks if don't need to support a variety of kernels that all use the same subclassed tensor, it only needs to be specific enough so we can dispatch to the correct kernel. maybe this can be simplified if we split the AQT in the future.
